base/pointer.jl

```diff
@@ -96,34 +96,168 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
             p::Ptr{T}, dims::NTuple{N,<:Integer}; own::Bool = false) where {T,N} =
     unsafe_wrap(Atype, p, convert(Tuple{Vararg{Int}}, dims), own = own)
 
+const ATOMIC_PTR_DOCS = """
+
+"""
+
 """
     unsafe_load(p::Ptr{T}, i::Integer=1)
+    unsafe_load(p::Ptr{T}, order::Symbol)
+    unsafe_load(p::Ptr{T}, i::Integer, order::Symbol)
 
 Load a value of type `T` from the address of the `i`th element (1-indexed) starting at `p`.
-This is equivalent to the C expression `p[i-1]`.
+This is equivalent to the C expression `p[i-1]`. Optionally, an ordering can be provided.
 
 The `unsafe` prefix on this function indicates that no validation is performed on the
 pointer `p` to ensure that it is valid. Like C, the programmer is responsible for ensuring
 that referenced memory is not freed or garbage collected while invoking this function.
 Incorrect usage may segfault your program or return garbage answers. Unlike C, dereferencing
 memory region allocated as different type may be valid provided that the types are compatible.
+
+This is compatible with loading pointers declared with `_Atomic` and `std::atomic`
```

Comment:
This clause should probably state that this only applies if `order` is given [, or if order != :not_atomic]. I forgot this is the text for both cases.

We also should try to make a decision whether `order != :not_atomic` should behave exactly the same as `unsafe_load`. Currently internally those may make slightly different alignment and size decisions. In particular, `aligned_sizeof` assumes the object is being stored in an Array. But here we are working with Ptr, and usually `unsafe_load`/`unsafe_store!` ignore mutability and the like and simply load or store bytes for an object of the given size at the given offset and allocate a *new* mutable object with those.

Reply:
If one doesn't actually spend anytime looking at internal code here I'd bet the assumption is that `:not_atomic` would perform that exact same as the unannotated method variant. Is there an important reason we have `:not_atomic` behave differently?

Reply:
No, they probably should behave the same as the non atomic versions. There are different behaviors of those however, so the code needs to be updated accordingly to make sure it does that.

Reply:
Sorry, that's obvious now that I'm reading through these comments again. Should this PR change the generated code and runtime C code for this or solve this in a separate PR?

Reply:
It should be a separate PR